### PR TITLE
Adapter-native Request and Response (enables full Cloudflare Worker functionality)

### PR DESCRIPTION
**Pull request by skirsten**
### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [ ] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpx changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

---

Hi, this is my weekend project that allows to directly interact with the Request and Response objects (and more) used by the adapter (if implemented).
This makes most sense for the two Cloudflare adapters, as it exposes the full CF Worker functionality:
- Env vars
- KV Namespaces
- waitUntil
- the Request.cf and Response.cf property
- The ability to proxy requests through without buffering
- *basically anything a normal worker has access to*

Here is a example that tests some of the Cloudflare specific functionality exposed by the adapter:
```ts
import type { AdapterRequestHandler } from '@sveltejs/adapter-cloudflare';

export const get: AdapterRequestHandler = async (request) => {
  if (request.adapter) {
    const colo = request.adapter.request.cf?.colo;

    request.adapter.ctx.waitUntil(
      (async () => {
        // Do stuff here
        return;
      })()
    );

    return {
      adapter: {
        // No buffering in memory happens here! Its passed straight through.
        response: fetch(
          `https://fakeimg.pl/440x230/282828/eae0d0/?retina=1&text=${encodeURIComponent(
            `proxied by CF in ${colo}`
          )}`,
          { cf: { cacheEverything: true } }
        )
      }
    };
  }
  return {
    status: 302,
    headers: {
      location: `https://fakeimg.pl/440x230/282828/eae0d0/?retina=1&text=${encodeURIComponent(
        'fallback to redirect'
      )}`
    }
  };
};
```

To test it, I build a example project https://github.com/skirsten/kit/tree/master, deployed here https://svelte-kit-cf-adapter-demo.pages.dev/.

It should be 100% backwards compatible and there is no need to use this new feature in your project.
This closes a number of issues:
- https://github.com/sveltejs/kit/issues/2844
- https://github.com/sveltejs/kit/issues/2807
- https://github.com/sveltejs/kit/issues/2606
- *and probably more*

Unfortunately I don't have the time to finish this, so this is a draft. If the approach taken here is okay with the maintainers, I would be glad if somebody can take this over and finish it.

### Missing
- [ ] Mismatch between Cloudflare Worker types and other types (Request, fetch, etc.). I am stuck here, I have no idea how to force it to use the Cloudflare Worker types...
- [ ] Only tested `adapter-cloudflare` not the `adapter-cloudflare-workers` and not `adapter-netlify` (was only a proof of concept to keep it generic)
- [ ] Update docs
- [ ] Test (automated?) all the edge-case interactions e.g. pre-rendered, handle hook, etc.
- [ ] Fix one failing CI. Seems unrelated.

### CODE REVIEW

File: examples/hn.svelte.dev/src/routes/__error.svelte [hunk 1]
```diff
@@ -1,4 +1,5 @@
 <script context="module">
+	/** @type {import('@sveltejs/kit').ErrorLoad} */
```
Comment (benmccann) [inline examples/hn.svelte.dev/src/routes/__error.svelte:2]: this is kind of a big PR. what do you think about sending the changes to `hn.svelte.dev` in a separate PR to break things up a bit? we could merge that change more quickly

**Comment by Rich-Harris** [comment]
As of #3384, the platform `Request` object is passed to your SvelteKit app, which IIUC enables what you need. The wrinkle, of course, is that you can't use the extra bits during development — we may in future need to expose a way for adapters to augment the request.

Anyway I'll close this — if there's more stuff that needs to happen around this, by all means raise a new issue. Thanks!

**Comment by tylerbrostrom** [comment]
@Rich-Harris Heads up: #3384 doesn’t (fully) resolve the issues that would be resolved by this PR.

#3426 lays out how those issues might be resolved, taking #3384 into account.
